File: Java/845.java
__________________________________________________________________________________________________
sample 1 ms submission
class Solution {

    public int longestMountain(int[] A) {
        int n;
        if (A == null || (n = A.length) < 3) {
            return 0;
        }
        int i = 1, max = 0;
        while (i < n) {
            // find the start of the up left mountain foot
            while (i < n && A[i] <= A[i - 1]) {
                i++;
            }
            int start = i - 1;
            while (i < n && A[i] > A[i - 1]) {
                i++;
            }
            if (i >= n) {
                return max;
            }
            if (A[i] == A[i - 1]) {
                continue;
            }
            while (i < n && A[i] < A[i - 1]) {
                i++;
            }
            int end = i - 1;
            max = Math.max(max, end - start + 1);
        }
        return max;
    }
}
__________________________________________________________________________________________________
sample 38832 kb submission
class Solution {
    public int longestMountain(int[] A) {
        int N = A.length;
        int count = 0;
        int j = 0;
        
        while(j < N){
            int end = j;
            
            if(end+1 < N && A[end] < A[end+1]){
                while(end+1 < N && A[end] < A[end+1]){
                    end++;
                }
                
                if(end+1 < N && A[end] > A[end+1]){
                    while(end + 1 < N && A[end] > A[end+1]){
                        end++;
                    }
                    count = Math.max(count, end-j+1);
                }
            }
            j = Math.max(end, j+1);
        }
        return count;
    }
}
__________________________________________________________________________________________________

File: Python3/845.py
__________________________________________________________________________________________________
sample 176 ms submission
class Solution:
    def longestMountain(self, A: List[int]) -> int:
        if len(A) <= 2:
            return 0
        
        is_asc = True
        has_asc = False
        has_desc = False
        prev = A[0]
        max_len = 0
        length = 0
        
        for cur in A[1:]:
            # "Downhill"
            if prev > cur:
                # Reached the peak, going down
                if is_asc and has_asc:
                    has_desc = True
                    is_asc = False
                    has_asc = False
                    length += 1

                    # Rolling down the mountain
                if not is_asc:
                    length += 1
            
            # "Uphill"
            if cur > prev:
                # First step in mountain
                if is_asc:
                    has_asc = True
                    length += 1
                # previously downhill, encountered another uphill
                else:
                    max_len = max(length, max_len)
                    length = 1
                    is_asc = True
                    has_asc = True
            
            if cur == prev:
                if not is_asc:
                    max_len = max(length, max_len)
                length = 0
                is_asc = True
                has_asc = False
                
            prev = cur
        
        if not has_desc:
            return 0
        
        return max(max_len, length)
__________________________________________________________________________________________________
sample 180 ms submission
class Solution:
    def longestMountain(self, A: List[int]) -> int:
        if len(A) < 3:
            return 0
        
        ans = 0
        length = 1
        temp = A[0]
        climbUp = temp < A[1]
        print(temp, A[1], climbUp, length)
        for x in A[1:]:
            if climbUp:
                length += 1
                if not x > temp:
                    climbUp = False
                    if x == temp:
                        length = 1
            elif length == 1:
                if x > temp:
                    climbUp = True
                    length += 1
            else:
                if x < temp:
                    length += 1
                else:
                    ans = max(length, ans)
                    length = 1
                    if x > temp:
                        climbUp = True
                        length += 1
            
            # print(temp, x, climbUp, length)
            temp = x
        
        if length > ans and not climbUp:
            ans = length
                    
        return ans if ans > 2 else 0
__________________________________________________________________________________________________
